lib/segment/src/vector_storage/quantized/quantized_multivector_storage.rs

```diff
@@ -0,0 +1,180 @@
+use std::fs::File;
+use std::io::{Read, Write};
+use std::marker::PhantomData;
+use std::path::Path;
+
+use common::types::{PointOffsetType, ScoreType};
+use quantization::{EncodedVectors, VectorParameters};
+use serde::{Deserialize, Serialize};
+
+use crate::common::operation_error::{OperationError, OperationResult};
+use crate::data_types::vectors::TypedMultiDenseVectorRef;
+use crate::types::{MultiVectorComparator, MultiVectorConfig};
+
+#[derive(Clone, Copy, Debug, Default, Serialize, Deserialize, PartialEq)]
+pub struct MultivectorOffset {
+    pub offset: PointOffsetType,
+    pub count: PointOffsetType,
+}
+
+pub struct QuantizedMultivectorStorage<TEncodedQuery, QuantizedStorage>
+where
+    TEncodedQuery: Sized,
+    QuantizedStorage: EncodedVectors<TEncodedQuery>,
+{
+    quantized_storage: QuantizedStorage,
+    offsets: Vec<MultivectorOffset>,
+    dim: usize,
+    multi_vector_config: MultiVectorConfig,
+    encoded_query: PhantomData<TEncodedQuery>,
+}
+
+impl<TEncodedQuery, QuantizedStorage> QuantizedMultivectorStorage<TEncodedQuery, QuantizedStorage>
+where
+    TEncodedQuery: Sized,
+    QuantizedStorage: EncodedVectors<TEncodedQuery>,
+{
+    pub fn new(
+        dim: usize,
+        quantized_storage: QuantizedStorage,
+        offsets: Vec<MultivectorOffset>,
+        multi_vector_config: MultiVectorConfig,
+    ) -> Self {
+        Self {
+            dim,
+            quantized_storage,
+            offsets,
+            multi_vector_config,
+            encoded_query: PhantomData,
+        }
+    }
+
+    pub fn save_multi(
+        &self,
+        data_path: &Path,
+        meta_path: &Path,
+        offsets_path: &Path,
+    ) -> OperationResult<()> {
+        let offsets_serialized = bincode::serialize(&self.offsets).map_err(|_| {
+            OperationError::service_error("Cannot serialize quantized multivector offsets")
+        })?;
+        let mut file = File::create(offsets_path)?;
+        file.write_all(&offsets_serialized)?;
+        file.flush()?;
+
+        Ok(self.quantized_storage.save(data_path, meta_path)?)
+    }
+
+    pub fn load_multi(
+        data_path: &Path,
+        meta_path: &Path,
+        offsets_path: &Path,
+        vector_parameters: &VectorParameters,
+        multi_vector_config: &MultiVectorConfig,
+    ) -> OperationResult<Self> {
+        let mut file = File::open(offsets_path)?;
+        let mut offsets_serialized = Vec::new();
+        file.read_to_end(&mut offsets_serialized)?;
+        let offsets = bincode::deserialize(&offsets_serialized).map_err(|_| {
```

Comment:
same comment about logging

Reply:
I guess the error message will be shown in logs anyway. I added specific filename into error message